client/src/lib/dependency-injection/inject.js

```diff
@@ -9,66 +11,74 @@ import injectorContext from './injectorContext';
+    if (dependencies && !Array.isArray(dependencies)) {
+      throw new Error(`
       withInjector() passed an argument for dependencies that is ${typeof deps}. 
       Must be a string or array of named dependencies.
     `);
-  }
+    }
 
-  if (mapDependenciesToProps && typeof mapDependenciesToProps !== 'function') {
-    throw new Error(`
+    if (mapDependenciesToProps && typeof mapDependenciesToProps !== 'function') {
+      throw new Error(`
       Second parameter of inject() [mapDependenciesToProps] must be a function, taking the resolved
       dependencies as enumerated arguments, and returning a map of prop names to dependencies.
     `);
-  }
+    }
 
-  if (getContext && typeof getContext !== 'function') {
-    throw new Error(`
+    if (getContext && typeof getContext !== 'function') {
+      throw new Error(`
       Third parameter of inject() [getContext] must be a function, taking the component's props
       as the single parameter, and returning a string representing the Injector
       context to use throughout the component.
     `);
-  }
+    }
 
-  // eslint-disable-next-line react/prefer-stateless-function
-  class Injected extends React.Component {
-    render() {
-      let props = {};
-      const context = getContext ? getContext(this.props) : null;
-      if (dependencies) {
-        const resolved = dependencies.map(dep => this.context.injector.get(dep, context));
+    // eslint-disable-next-line react/prefer-stateless-function
+    class Injector extends React.Component {
+      render() {
+        let props = {};
+        const context = getContext ? getContext(this.props) : null;
+        if (dependencies) {
+          const resolved = dependencies.map(dep => this.context.injector.get(dep, context));
 
-        if (mapDependenciesToProps) {
-          props = mapDependenciesToProps(...resolved);
-        } else {
-          // If no mapping function is given, mirror the prop names and dependency names
-          dependencies.forEach((dep, index) => {
-            props[dep] = resolved[index];
-          });
-        }
+          if (mapDependenciesToProps) {
+            props = mapDependenciesToProps(...resolved);
+          } else {
+            // If no mapping function is given, mirror the prop names and dependency names
+            dependencies.forEach((dep, index) => {
+              props[dep] = resolved[index];
+            });
+          }
 
-        if (!props || typeof props !== 'object') {
-          throw new Error(`
+          if (!props || typeof props !== 'object') {
+            throw new Error(`
             mapDepedenciesToProps parameter passed to inject()
             should return an object that maps prop names to dependencies
           `);
+          }
         }
+        const newProps = {
+          ...props,
+          ...this.props,
+        };
+        return <Component {...newProps} />;
       }
-      const newProps = {
-        ...this.props,
-        ...props,
-      };
```

Comment:
The other option is to allow the injected component props be replaceable by the calling function/component.

An example of UploadField:
`<UploadField UploadFieldItem={CustomComponent} {...props} />`
would allow for customising the `UploadFieldItem` child component which is injected to `UploadField`

I have a separate pull request coming up which switches `UploadFieldItem` from `import` to using `inject()`

Reply:
I still think this should be offered as a function, which does this by default. There may be cases where you want the local props to be overriden by the injector props, or better yet, create, alter, or remove props based on the diff.

Reply:
yeah, but I struggled to find a case for that... considering you're in the parent context and should have most of the props that are being passed down to the injected component?

I also couldn't think of how to provide that function between `inject()` and the `Component`. The `inject()` has the second param `mapDependenciesToProps` - but this is instance it is inaccessible and already executed.

Reply:
I also think that if the parent enforces a subcomponent via props, it should not be overridden by `inject`. Maybe the inject HOC could check if the prop is already set and leave it untouched if it is?

This would give developers the possibility to change sub-components based on state, which could be useful with something like a sortable-uploadfield, where sorting can be enabled via a prop.

Reply:
Hm, I like that idea. If I'm interpreting this right, then it would work for all cases. Eg. to globally replace a component, replace it for a given context, or even for a context that is generated dynamically (eg. based on some props). Sounds good.

Reply:
Yeah, I like it. I think there's a really good use case for passing the context to children, for all the reasons @bummzack has stated.

This is getting a bit meta, but rather than passing the context as a prop, what about passing it via `context`? That way you're not bound to the parent-child hierarchy.  You could in theory leave `UploadFieldItem` alone, but customise `UploadFieldItemThumbnail` (or whatever) based on the `UploadField` context.

Reply:
Passing the context in the context… down the rabbit hole we go 👍 
Is it this context we're talking about here? https://reactjs.org/docs/context.html#why-not-to-use-context
Seems like they would rather un-invent that feature :)
I'm happy with either approach. If context is already being used for SS react-components this way, we can surely expand on that.

Reply:
Yeah, it's not an API that's often viewed in a very positive light, and that's unfortunate. The point of that article (which I find a bit too absolute and admonishing) is that the basic design pattern of passing props should not be broken without serious consideration. That doesn't mean that it's never a good idea. Sometimes it's exactly the right idea. It's useful anytime you're trying to influence how a component behaves based on _where_ it is, rather than _how_ it is. Drag-and-drop libraries, redux stores, even my [react-selectable](https://github.com/unclecheese/react-selectable) library makes good use of it.

The SS React injector itself is a context service, and I think that's the best argument for using context here.

Reply:
I had bounced the idea of using `react.context` for this, it definitely suits the case well.

I had also wondered if it would play nice with the `context` tweaking I mentioned above, you want to change the context in the middle for a new transform - or if that's even a necessary feature?
Simplicity and I know that it'd work with props were the reasons why I've implemented it this way :)

Actually, not allowing context tweaking could work if `inject` didn't override Component props provided by parent... and simplify the API/magic a bit :)

Reply:
Guys, is there some documentation available for this? I'm struggling with this use case:
 - `UploadField` uses `UploadFieldItem` as child component.
 - I decorate `UploadField` with my own HOC
 - Depending on a prop passed to my HOC, I want to decorate `UploadFieldItem` with a HOC as well.

Is that even possible?

Reply:
Hey @bummzack, you would be better off to have the "decorate with HOC" logic flipped, so the HOC will "pass-through" if certain props don't meet a condition.
So your UploadFieldItem is decorated with the HOC without knowing if the decorator is active or not.

`fieldHolder` in `silverstripe-admin` is a good example of passing-through if the "noHolder" prop was turned on.

HOCs shouldn't be hot-loaded on render, as it leads to performance hits (noticeably if it has many children)
It's a good idea to define constants for the React render engine to use - since it does a shallow reference check on the class/function that it tries to render to see if it should re-render for changes.

Reply:
@flamerohr Thanks for the explanation

I actually don't need hot-loading. It's for the use-case where the `UploadField` doesn't need any sorting at all, so I don't really want my HOC and styles to be applied at all. I'm not really sure how I can achieve that, since I'm decorating the `UploadField`, but I actually only need it to be decorated when the user makes use of the matching PHP field class (`SortableUploadField` instead of `UploadField`).

That's the first issue and it could be solved by a pattern like used in `fieldHolder`, where you just render the "base" component if there was no `sortable` property being passed from the PHP field.

The second issue is what I outlined above with the `UploadFieldItem`. If I don't decorate the parent, I obviously also don't want to decorate the child. But since I'm unable to pass properties to the child, I don't know how I would solve that?

Reply:
yeah, with the current API, it's a convoluted way of doing it...

since you're decorating the UploadField, you could try take advantage of the new-ish API where you could overwrite an injected component by passing your own component to the parent via props.

It was a sort of abuse of that API, but you can do something like:
```
constructor(props) {
  super(props);

  this.DecoratedItem = itemDecorator(props.UploadFieldItem);
}

render() {
  const Item = this.props.sortable
    ? this.DecoratedItem
    : props.UploadFieldItem;

  return <UploadField UploadFieldItem={Item} />;
}
```

I suggest constructing the decorated item in the constructor so you're not hot-loading the decorated item every render.
I think the main caveat with this approach is if another decorator before your decorator did the same thing, then it could become really confusing...
maybe we could investigate exposing a `getItemProps` prop callback, so then item props are extensible and this way we can go without the default over-exposing props :)

Reply:
Thanks, I'll give this a try! Exposing `getItemProps` sounds like a good idea now indeed… and we have gone full circle :)